vault_service/views/utils.py

```diff
@@ -85,7 +91,483 @@ def check_request(request):
+
+    query: qterm
+    
+    qterm: anyterm -> qterm | phrase | PREPEND -> prepend
+    
+    PREPEND.2: /=/ | /\+/ | /\-/
+        
+    phrase: DOUBLE_QUOTED_STRING | SINGLE_QUOTED_STRING
+
+    DOUBLE_QUOTED_STRING.3  : /"[^"]*"/ | /\+"[^"]*"/ | /\-"[^"]*"/
+    SINGLE_QUOTED_STRING.3  : /'[^']*'/ | /\+'[^']*'/ | /\-'[^']*'/
+    
+    anyterm: /[^)^\] \(^\n^\r]+/
+    
+    operator: OPERATOR | NEWLINE
+
+    OPERATOR.2: "and" | "AND" | "or" | "OR" | "not" | "NOT" | "AND NOT" | "and not" | /,/ 
+
+    %import common.LETTER
+    %import common.ESCAPED_STRING
+    %import common.FLOAT
+    %import common.DIGIT
+    %import common.WS_INLINE
+    %import common.NEWLINE
+    
+    %ignore WS_INLINE
+    
+    """, parser="lalr")
+
+    tree = grammar.parse(data)
+
+    return tree
+
+
+class TreeVisitor(Visitor):
+    """
+    Visitor class to transform the parsed tree into a BBB-style query.
+    The final constructed query is stored in v.visit(tree).output
+    """
+    def start(self, node):
+        out = []
+        for x in node.children:
+            if hasattr(x, 'output'):
+                out.append(getattr(x, 'output'))
+            else:
+                pass
+        tmp = ' '.join(out)
+
+        node.output = tmp
+
+    def clause(self, node):
+        out = []
+        ops = ['and', 'AND', 'or', 'OR', 'not', 'NOT', 'and not', 'AND NOT']
+        prepend = ['=', '+', '-']
+        for x in node.children:
+            if hasattr(x, 'output'):
+                out.append(getattr(x, 'output'))
+        output = []
+        i = 0
+        for o in out:
+            if i == 0:
+                output.append(o)
+            else:
+                if output[i-1] in ops:
+                    output.append(o)
+                elif output[i-1] in prepend:
+                    if i < 2:
+                        output[i-1] += o
+                    else:
+                        output[i-1] = 'OR ' + output[i-1] + o
+                    output.append('')
+                elif o in ops or o in prepend:
+                    output.append(o)
+                else:
+                    output.append('OR ' + o)
+            i += 1
+
+        output = [x for x in output if x != '']
+
+        if len(output) > 1:
+            node.output = "({0})".format(' '.join(output))
+        else:
+            node.output = "{0}".format(' '.join(output))
+
+    def query(self, node):
+        node.output = node.children[0].output
+
+    def qterm(self, node):
+        node.output = node.children[0].output
+
+    def anyterm(self, node):
+        node.output = '{0}'.format(node.children[0].value.replace("'", "\'").replace('"', '\"').strip())
+
+    def phrase(self, node):
+        node.output = node.children[0].value.strip()
+
+    def prepend(self, node):
+        node.output = node.children[0].value.strip()
+
+    def operator(self, node):
+        v = node.children[0].value
+        if v not in ['and', 'AND', 'or', 'OR', 'not', 'NOT', 'and not', 'AND NOT']:
```

Comment:
i'd recommend something like 
`
v = node.children[0].value.upper()
if v not in ['AND', 'OR', 'NOT', 'AND NOT']...`